File: csharp/1504.count-submatrices-with-all-ones.cs
/*
 * @lc app=leetcode id=1504 lang=csharp
 *
 * [1504] Count Submatrices With All Ones
 */

using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {
    public int NumSubmat(int[][] mat) {
        int m = mat.Length, n = mat[0].Length, ret = 0;
        int[] up = new int[n];
        var stack = new Stack<int>();
        for (int i = 0; i < m; ++i) {
            int[] sums = new int[n];
            for (int j = 0; j < n; ++j) {
                if (mat[i][j] == 0) {
                    up[j] = 0;
                    stack.Clear();
                    stack.Push(j);
                    continue;
                }
                ++up[j];
                while (stack.TryPeek(out int k) && up[k] >= up[j]) {
                    stack.Pop();
                }
                if (stack.TryPeek(out int p)) {
                    sums[j] = sums[p] + up[j] * (j - p);
                } else {
                    sums[j] = up[j] * (j + 1);
                }
                stack.Push(j);
            }
            ret += sums.Sum();
            stack.Clear();
        }
        return ret;
    }
}
// @lc code=end

File: java/1504.count-submatrices-with-all-ones.java
/*
 * @lc app=leetcode id=1504 lang=java
 *
 * [1504] Count Submatrices With All Ones
 */

import java.util.Stack;

// @lc code=start
class Solution {
    public int numSubmat(int[][] mat) {
        int m = mat.length, n = mat[0].length, ret = 0;
        int[][] sums = new int[2][n];
        int[] upCount = new int[n];
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < m; ++i) {
            int[] sum = sums[i % 2];
            int[] row = mat[i];
            stack.clear();
            for (int j = 0; j < n; ++j) {
                if (row[j] == 0) {
                    stack.clear();
                    stack.push(j);
                    upCount[j] = 0;
                    sum[j] = 0;
                    continue;
                }
                ++upCount[j];
                while (!stack.empty() && upCount[stack.peek()] >= upCount[j]) {
                    stack.pop();
                }
                if (stack.empty()) {
                    sum[j] = upCount[j] * (j + 1);
                } else {
                    int p = stack.peek();
                    sum[j] = sum[p] + upCount[j] * (j - p);
                }
                ret += sum[j];
                stack.push(j);
            }
        }
        return ret;
    }
}
// @lc code=end

